Implement hcp packer data source.

SwampDragons opened the pull request:
This is a rough copy of the terraform provider's data source, to make beta user experience consistent across both sources

I had to re-implement the model configs in the datasource, to make sure that we can generate the appropriate hcl structures properly. Otherwise, the logic and structure of the returned value should match what we have in the terraform provider. 

usage example: 

```
data "packer-image-iteration" "hardened-source" {
  bucket_name = "hardened-ubuntu-16-04"
  channel = "megan-test"
}

local "packer-registry-source" {
  expression = [ for image in flatten(data.packer-image-iteration.hardened-source.builds[*].images[*]): image.image_id ][0]
}

source "amazon-ebs" "packer-secondary" {
  source_ami       = local.packer-registry-source
  ...
}

```

azr commented on datasource/packer-image-iteration/data.go:
```diff
@@ -0,0 +1,164 @@
+	return nil
+}
+
+// Copy of []*models.HashicorpCloudPackerBuild. Need to copy so we can generate
+// the HCL spec.
+type ParBuild struct {
```
I wonder if type aliasing would work here 🤔 : 

```suggestion
// Alias of models.HashicorpCloudPackerBuild. So we can generate the HCL spec.
type ParBuild = models.HashicorpCloudPackerBuild
```

azr replied:
(to not have to update the field of this struct)

azr commented on datasource/packer-image-iteration/data.go:
```diff
@@ -0,0 +1,164 @@
+	return nil
+}
+
+// Copy of []*models.HashicorpCloudPackerBuild. Need to copy so we can generate
+// the HCL spec.
+type ParBuild struct {
+	// aws
+	CloudProvider string `mapstructure:"cloud_provider"`
+	// builder or post-processor used to build this
+	ComponentType string `mapstructure:"component_type"`
+	// created at
+	// Format: date-time
+	CreatedAt string `mapstructure:"created_at"`
+	// ULID
+	ID string `mapstructure:"id"`
+	// images
+	Images []ParImage `mapstructure:"images"`
+	// ULID of the iteration
+	IterationID string `mapstructure:"iteration_id"`
+	// unstructured metadata
+	Labels map[string]string `mapstructure:"labels"`
+	// packer run uuid
+	PackerRunUUID string `mapstructure:"packer_run_uuid"`
+	// complete
+	Status string `mapstructure:"status"`
+	// updated at
+	// Format: date-time
+	UpdatedAt string `mapstructure:"updated_at"`
+}
```
I wonder if type aliasing would work here 🤔 : 

```suggestion
// Alias of models.HashicorpCloudPackerBuild. So we can generate the HCL spec.
type ParBuild = models.HashicorpCloudPackerBuild
```

(to not have to maintain the struct)

sylviamoss replied:
It probably doesn't work because of the same problem Megan was having before, the pointer. For that, I think we would have to make the mapstructure-to-hcl translate better pointers.

SwampDragons replied:
Correct. This was my first try.

azr reviewed (approved):
Super nice, LGTM, I think we are just missing a docs page.

sylviamoss commented on datasource/packer-image-iteration/data.go:
```diff
@@ -0,0 +1,164 @@
+	Region string `mapstructure:"region,omitempty"`
+}
+
+type DatasourceOutput struct {
```
If you write docs here it will create the mdx file for this too.

sylviamoss commented on datasource/packer-image-iteration/data.go:
```diff
@@ -0,0 +1,164 @@
+	Labels map[string]string `mapstructure:"labels"`
+	// packer run uuid
+	PackerRunUUID string `mapstructure:"packer_run_uuid"`
+	// complete
```
If we use the output docs, I think this one needs a better documentation

SwampDragons replied:
ok!

sylviamoss replied:
Thanks!!

sylviamoss reviewed (approved):
LGTM 👍🏼 
This one is very straightforward, I haven't tested it myself.

SwampDragons commented:
ok, I updated the docs and added a docs page, but I didn't actually add the page to the website navigation yet because we need to wait until the HCP Packer registry enters public beta.
